### PR TITLE
saving time on file, in case of a crash

### PR DESCRIPTION
**Pull request by taltaub22**

### CODE REVIEW

**Comment by taltaub22** [comment]
1. Will fix
4. Yes, my bad

**Comment by taltaub22** [comment]
@idanstark42 Fixed.
We discussed on the call to add 5 seconds so teams will not get affected by this error.
If you think otherwise lets talk about it.
@alanggreen @yosikarl

**Comment by idanstark42** [comment]
1. If we add this 5 seconds it's something that we shold tell the Tournament organizers. It could look worse then the crushes. The crush is the system missbehaving, but when adding a random 0-5 seconds to a match, it could make the competition unfare from the POV of the teams. @alanggreen your tohught on this point?
2. You forgot to change the functions names in the manager

**Comment by alanggreen** [comment]
The idea of the extra time was this:
If the clock crashes somewhere in the middle of the match, we resume the countdown when it comes up.
But what happens if the clock crashes 1 sec before end? By the time clock restarts, the 2:30 is over there won't be an end sound.
Solution: Up till x secs after 2:30, if the time file is still there when clock restarts, play the "end" before deleting time file.

File: server/time-saver.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+const FILE_NAME = 'timer-start-time.json'
+
+exports.TimeSaver = class {
+  writeTimeToFile (time) {
+    return writeFile(path.join(DATA_DIR, FILE_NAME), time.getTime())
+  }
+
+  getTimeFromFile () {
+    return readFile(path.join(DATA_DIR, FILE_NAME), 'utf-8').then(time => {
+      time = JSON.parse(time)
```
Comment (alanggreen) [inline server/time-saver.js]: JSON.parse can generate a [Syntax Error exception](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/parse#Exceptions). Suggest to catch and handle, including log.warn

**Comment by idanstark42** [comment]
Please get @2roy999's opinion before you merge this

File: server/time-saver.js [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+      if (time) {
+        return new Date(parseInt(time))
```
Comment (alanggreen) [inline server/time-saver.js]: If time is not valid then date will return Invalid Date

File: server/manager.js [hunk 3]
```diff
@@ -69,6 +72,23 @@ exports.ClockManager = class extends EventEmitter {
...
+        const timeRun = Math.floor((currentTime.getTime() - time.getTime()) / 1000)
+        const matchTimeWithBuffer = MATCH_TIME + MATCH_TIME_BUFFER
+
+        if (timeRun < matchTimeWithBuffer) {
+          const timeLeft = matchTimeWithBuffer - timeRun
+          this._clock.startCountdown(timeLeft)
+        } else {
+          this._clock.startCountdown(0)
+        }
+      }
```
Comment (alanggreen) [inline server/manager.js]: Unless ```getTimeFromFile``` handles exceptions, maybe you want a .catch() here

File: server/manager.js [hunk 4]
```diff
@@ -69,6 +72,23 @@ exports.ClockManager = class extends EventEmitter {
...
 
+    this._timeSaver.getTimeFromFile().then(time => {
+      if (time) {
+        this._status = RUNNING
+
+        const currentTime = new Date()
+        const timeRun = Math.floor((currentTime.getTime() - time.getTime()) / 1000)
+        const matchTimeWithBuffer = MATCH_TIME + MATCH_TIME_BUFFER
+
+        if (timeRun < matchTimeWithBuffer) {
```
Comment (alanggreen) [inline server/manager.js]: Does this code implement this?
>The idea of the extra time was this:
If the clock crashes somewhere in the middle of the match, we resume the countdown when it comes up.
But what happens if the clock crashes 1 sec before end? By the time clock restarts, the 2:30 is over there won't be an end sound.
Solution: Up till x secs after 2:30, if the time file is still there when clock restarts, play the "end" before deleting time file.

**Pull request merged**